### PR TITLE
For discussion: Alignments

### PR DESCRIPTION
**Pull request by jkbelcher**
This is a snowball at the top of a big hill, so I'm not pushing it far before checking in to see if you're interested in the direction.

I've been building UI elements recently and would love to gain the efficiency of fill/stretch alignments.  Example use case: after adding a [newVerticalContainer] column, setting it to non-standard width 70, I'd rather not call `setWidth(70)` on every child.  Ideally it would look like:

```
// Use fill for secondary direction
UI2dContainer column = newVerticalContainer(70);
column.setContentHAlignm(HAlign.FILL);  // Default secondary alignment for children
column.addChild(newLabel("Autosized!"));
column.addChild(newKnob(param).setHAlign(HAlign.CENTER));  // Child HAlign takes priority
```

The `FillPercent` variables mimic the behavior I used on the TouchOSC project last winter.  It allows a child to say: Size me to X percent of the space remaining after fixed-width children are placed.  In a lot of cases this would be one child set to fill 100% of remaining space.  But in some cases such as transition controls I might do:
```
// Use fill for primary direction
transitionControlRow.setContentHAlign(FILL);
transitionControlRow.addChildren(
  newButton(enabled, 15),  // 15 pixels wide
  newDropDown(blendMode).setHFillPercent(70),
  newTimeBox(transitionTime).setHFillPercent(30)
);

```

### CODE REVIEW

**Comment by mcslee** [comment]
Urgh, yeah I'm inclined to avoid pushing this snowball down this hill. I think advanced layout management probably wants to live outside of `UI2dContainer` or we're gonna risk a massive pixel-counting-cruft-code bloat growing and growing inside there, which feels not ideal.

I was just finally implementing drag and drop and found myself thinking `HORIZONTAL_EVEN` and `HORIZONTAL` shouldn't be separate modes, but rather there should just be a separate flag to `setChildSpacing()` that specifies if they're also sized.

For sizing in these columns, I think a lot of that could be done with smarter column helpers, e.g. looking at the core loop in `addColumn`

```
     for (UI2dComponent component : components) {
        component.addToContainer(col);
      }
```

There's no reason this couldn't `component.setWidth(colWidth);` for instance, or do other sizing behaviors based upon more complex arguments to the `addColumn` helper. Or a dynamic column helper class could do this.

Basically, I think these use cases can be accomplished without having to add new attributes on the `UI2dComponent` class or new layout managers. Seem reasonable?

**Comment by jkbelcher** [comment]
> I think advanced layout management probably wants to live outside of UI2dContainer or we're gonna risk a massive pixel-counting-cruft-code bloat growing and growing inside there, which feels not ideal.

Agreed, advanced layout management could live in child classes or extend an inner class (or helpers?), rather than accumulate in a long list of `else if`s in the `reflow()` method.  That would make it more extensible for user-defined layout logic too.

> thinking HORIZONTAL_EVEN and HORIZONTAL shouldn't be separate modes, but rather there should just be a separate flag to setChildSpacing() that specifies if they're also sized.

Seems reasonable. 

`HORIZONTAL_GRID` feels incorrectly named... it's effectively just adding a wrap() function to `HORIZONTAL`, right?  The only reason it would look like a grid is if all the children are the same width.  Which is not a restraint currently enforced by the mode.  To me a grid is a fixed number of rows and columns, with each child's location being a cell with an address [row, column].

>      for (UI2dComponent component : components) {
>         component.addToContainer(col);
>       }
> There's no reason this couldn't component.setWidth(colWidth); for instance

Fine with me if you can handle all the use cases.  I have two main use cases that inspired this:
1) A column defaulting to stretch children, but one of the children is fixed width 50 and centered.
2) A modulator UI with a three columns stacked against the left, and then two columns stacked against the right.  TBH doing this feels clunky:
`newVerticalThing.setWidth(12).setX(uiModulator.getContentWidth() - 12)`

This feels like taking the long way around... the goal is to place a child by right-offset, but instead we're retrieving the parent width, subtracting the child width [and the width of the stuff next to it], and then placing the child by left-offset.  It should be a simple as addChild(child, RIGHT).

It feels like there's room for an alignment enum in here.  And a big gain would be: doing high-level resize with the full child hierarchy reflowing to the intended relative sizes & placements.

**Comment by mcslee** [comment]
Agreed grid is not really a grid either. One other thing worth mentioning here, the `Layout` stuff is only really needed when *dynamic* rearrangement of the children is desired automatically in response to children being added/removed/resized which triggers a `reflow` operation.

For static layouts that are just construct-once-and-use, helper methods to specify the positioning properties are totally sufficient.

I'll look at cleaning this stuff up at some point but quite a few things higher on my list just at the moment. Agreed that having some helper args on `addToContainer()` to automatically size-and-place while adding would be cool and save boilerplate repetition. I'll add those.

**Comment by jkbelcher** [comment]
> One other thing worth mentioning here, the Layout stuff is only really needed when dynamic rearrangement of the children is desired automatically in response to children being added/removed/resized which triggers a reflow operation.

I was imagining in addition to the collection changing, that a full window resize would trigger a `reflow` on all children.  At the moment there are 4 fixed-width/height areas, one tied to each edge of the window.  But hypothetically if there was a dialog that was slightly smaller than the full window... I could see an Alignment property being helpful in a reflow operation.

> Agreed that having some helper args on addToContainer() to automatically size-and-place while adding would be cool and save boilerplate repetition. I'll add those.

That sounds great.  Thanks.

**Comment by mcslee** [comment]
Okay @jkbelcher try this out:
https://github.com/mcslee/GLX/commit/61b9d7dabfc6f721c4692cf801af6fe84767c26b

There is a new enum UI2dContainer.Position. You can use this in a few ways.

```
// Method #1 - set the property upfront
child.setContainerPosition(UI2dContainer.Position.TOP_LEFT);
// The setting will take effect, and apply on reflows if there's no Layout set
container.addChildren(child, ...);

// Method #2 - set the property at call time, it will be preserved by
// reflows if there's no Layout set
child.addToContainer(container, UI2dContainer.Position.BOTTOM);

// Method #3 - do it afterwards
child.addToContainer(container);
// This will move the child and end up triggering a container redraw if it changes something
child.setContainerPosition(UI2dContainer.Position.LEFT);
```
 
I *think* this addresses most of your use cases. I did not add support here for mix-and-match between setting container positions and using a `UI2dContainer.Layout`. If there is a `Layout` set, then that takes precedence. But if it's `Layout.NONE` then these positions are preserved across resize/reflow operations.

Let me know how this works out for you.
